vpr/src/device/rr_spatial_lookup.cpp

```diff
@@ -132,6 +132,51 @@ std::vector<RRNodeId> RRSpatialLookup::find_channel_nodes(int x,
     return channel_nodes;
 }
 
+std::vector<RRNodeId> RRSpatialLookup::find_sink_nodes(int x,
+                                                       int y) const {
+    /* TODO: The implementation of this API should be worked 
+     * when rr_node_indices adapts RRNodeId natively!
+     */
+    std::vector<RRNodeId> sink_nodes;
+
+    /* Pre-check: the x, y, type are valid! Otherwise, return an empty vector */
+    if (x < 0 || y < 0) {
+        return sink_nodes;
+    }
+
+    VTR_ASSERT_SAFE(3 == rr_node_indices_[SINK].ndims());
+
+    /* Sanity check to ensure the x, y, side are in range 
+     * - Return a list of valid ids by searching in look-up when all the parameters are in range
+     * - Return an empty list if any out-of-range is detected
+     */
+    if (size_t(SINK) >= rr_node_indices_.size()) {
+        return sink_nodes;
+    }
+
+    if (size_t(x) >= rr_node_indices_[SINK].dim_size(0)) {
+        return sink_nodes;
+    }
+
+    if (size_t(y) >= rr_node_indices_[SINK].dim_size(1)) {
+        return sink_nodes;
+    }
+
+    /* By default, we always added the channel nodes to the TOP side (to save memory) */
```

Comment:
This comment looks wrong (talking about channels).  

This code also points out why we should rework the underlying rr_node_indices implementation; it's strange to have a multi-dimensional array matrix where some dimensions must always be accessed with a specific index.  Hopefully that is on your near-term agenda Xifan. Could put a TODO in to refactor that way, if you think such a comment would be a useful reminder.

Reply:
Sorry for the typo. Just fixed it.
I think it will be a good time to rethink the ``rr_node_indices`` data structure once all the related APIs are replaced. 
Actually, it is not too far. I expect only we need another 2 pull requests, before fully shadowing the current ``rr_node_indices``. 

I put a TODO to remind us during the refactoring. I will take a deep look into the ``rr_node_indices`` data structure and propose something.